src/js/heuristicblocking.js

```diff
@@ -104,33 +112,140 @@ HeuristicBlocker.prototype = {
-      // save the origin associated with the tab
-      log("Origin: " + origin + "\tURL: " + details.url);
       tabOrigins[details.tabId] = origin;
+      tabURLs[details.tabId] = details.url;
       return {};
     }
 
-    let tabOrigin = tabOrigins[details.tabId];
+    let tabOrigin = tabOrigins[details.tabId],
+      self = this;
 
     // ignore first-party requests
     if (!tabOrigin || origin == tabOrigin) {
       return {};
     }
 
     // abort if we already made a decision for this FQDN
-    let action = this.storage.getAction(fqdn);
+    let action = self.storage.getAction(fqdn);
     if (action != constants.NO_TRACKING && action != constants.ALLOW) {
       return {};
     }
 
-    // ignore if there are no tracking cookies
-    if (!hasCookieTracking(details, origin)) {
+    // check if there are tracking cookies
+    if (hasCookieTracking(details, origin)) {
+      self._recordPrevalence(fqdn, origin, tabOrigin);
       return {};
     }
 
-    this._recordPrevalence(fqdn, origin, tabOrigin);
+    if (checkForCookieShare && details.type == 'image') {
+      // get all cookies for the frame the request comes from and pass those to
+      // the cookie-share accounting function
+      chrome.cookies.getAll({
+        url: tabURLs[details.tabId]
+      }, function(cookies) {
+        self.pixelCookieShareAccounting(details, cookies);
+      });
+    }
+  },
+
+  /**
+   * Checks for cookie sharing: requests to third-party domains that include
+   * high entropy data from first-party cookies. Only catches plain-text
+   * verbatim sharing (b64 encoding + the like defeat it). Assumes any long
+   * string that doesn't contain URL fragments or stopwords is an identifier.
+   * Doesn't catch cookie syncing (3rd party -> 3rd party), but most of those
+   * tracking cookies should be blocked anyway.
+   *
+   * @param details are those from onBeforeSendHeaders
+   * @param cookies are the result of chrome.cookies.getAll()
+   * @returns {*}
+   */
+  pixelCookieShareAccounting: function (details, cookies) {
+    let fqdn = (new URI(details.url)).host,
+      origin = window.getBaseDomain(fqdn),
+      tabUrl = tabURLs[details.tabId],
+      tabOrigin = tabOrigins[details.tabId],
+      args = _extractArgs(details),
+      TRACKER_ENTROPY_THRESHOLD = 33,
+      MIN_STR_LEN = 8;
+
+    for (let key in args) {
+      if (!args.hasOwnProperty(key)) {
+        continue;
+      }
+
+      let value = args[key];
+
+      // the argument must be sufficiently long
+      if (!value || value.length < MIN_STR_LEN) {
+        continue;
+      }
+
+      // check if this argument is derived from a high-entropy first-party cookie
+      for (let cookie of cookies) {
+        if (cookie.domain == origin) {
+          continue;
+        }
+
+        // find the longest common substring between this arg and the cookies
+        // associated with the document
+        let substrings = utils.findCommonSubstrings(cookie.value, value) || [];
```

Comment:
Can we avoid calling `findCommonSubstrings()` if `cookie.value` is too short for us to care about? Any other cases for us to skip checking?